### PR TITLE
Fix cables status choice being provided by OPTIONS

### PR DESCRIPTION
**Pull request by FragmentedPacket**
### Fixes: #452 452

Remove the `metadata_class` as `StatusViewSetMixin` uses it to properly set the metadata returned.

### CODE REVIEW

**Comment by FragmentedPacket** [comment]
I would like to add some tests to make sure any model that uses the `Status` model will be tested that it's rendered metadata works and prevent any regression. Not sure if this should happen in two separate places (model or viewset tests and API tests) as the Ansible modules rely on this data.

**Comment by glennmatthews** [comment]
Looks like the `metadata_class` was added in Netbox https://github.com/netbox-community/netbox/issues/4695 to handle a different issue with the OPTIONS content on Cables - are we sure that removing this is the right thing to do, or do we need to instead extend with a custom `metadata_class` that supports both `status` and `termination_a_type` information?

**Comment by FragmentedPacket** [comment]
I thought I had responded, but I guess I didn't hit **comment**. I don't think I have an opinion on this so whichever way y'all think is beneficial and I will implement it.

**Comment by jathanism** [comment]
This should be fine see

> Looks like the `metadata_class` was added in Netbox [netbox-community/netbox#4695](https://github.com/netbox-community/netbox/issues/4695) to handle a different issue with the OPTIONS content on Cables - are we sure that removing this is the right thing to do, or do we need to instead extend with a custom `metadata_class` that supports both `status` and `termination_a_type` information?

If you look at the code, despite the filter conditions, the two metadata subclasses are doing nearly the same thing:
- `ContentTypeMetadata` is gating on field type `ContentTypeField`
- `StatusFieldMetadata` is gating on field type `StatusSerializerField`

I would reckon that with a little debugging we could probably adapt `StatusFieldMetadata` to do both in this case.

**Comment by FragmentedPacket** [comment]
I did try and get that working, but wasn't successful as I think that makes the most sense at this point. I will give it another go and hit you up if I have issues with it.

**Comment by FragmentedPacket** [comment]
I'm about to push up working changes as well as new tests, but wanted to get y'alls feedback on testing.

I think I also discovered another bug that the statuses that are returned are not properly filtered for the endpoint. Cables should only have three status choices and virtual machines should have 6. Numbers are arbitrary, but this is with a fresh install of Nautobot.

```python
>>> nb.dcim.cables.choices()['status']
[{'value': 'active', 'display': 'Active'}, {'value': 'available', 'display': 'Available'}, {'value': 'connected', 'display': 'Connected'}, {'value': 'container', 'display': 'Container'}, {'value': 'decommissioned', 'display': 'Decommissioned'}, {'value': 'decommissioning', 'display': 'Decommissioning'}, {'value': 'deprecated', 'display': 'Deprecated'}, {'value': 'deprovisioning', 'display': 'Deprovisioning'}, {'value': 'dhcp', 'display': 'DHCP'}, {'value': 'failed', 'display': 'Failed'}, {'value': 'inventory', 'display': 'Inventory'}, {'value': 'offline', 'display': 'Offline'}, {'value': 'planned', 'display': 'Planned'}, {'value': 'provisioning', 'display': 'Provisioning'}, {'value': 'reserved', 'display': 'Reserved'}, {'value': 'retired', 'display': 'Retired'}, {'value': 'slaac', 'display': 'SLAAC'}, {'value': 'staged', 'display': 'Staged'}, {'value': 'staging', 'display': 'Staging'}]
>>> nb.virtualization.virtual_machines.choices()['status']
[{'value': 'active', 'display': 'Active'}, {'value': 'available', 'display': 'Available'}, {'value': 'connected', 'display': 'Connected'}, {'value': 'container', 'display': 'Container'}, {'value': 'decommissioned', 'display': 'Decommissioned'}, {'value': 'decommissioning', 'display': 'Decommissioning'}, {'value': 'deprecated', 'display': 'Deprecated'}, {'value': 'deprovisioning', 'display': 'Deprovisioning'}, {'value': 'dhcp', 'display': 'DHCP'}, {'value': 'failed', 'display': 'Failed'}, {'value': 'inventory', 'display': 'Inventory'}, {'value': 'offline', 'display': 'Offline'}, {'value': 'planned', 'display': 'Planned'}, {'value': 'provisioning', 'display': 'Provisioning'}, {'value': 'reserved', 'display': 'Reserved'}, {'value': 'retired', 'display': 'Retired'}, {'value': 'slaac', 'display': 'SLAAC'}, {'value': 'staged', 'display': 'Staged'}, {'value': 'staging', 'display': 'Staging'}]
```

I haven't been able to dig into it too much, but I believe this is the culprit.

https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/api/serializers.py#L621

**Comment by glennmatthews** [comment]
I think the bug identification and your analysis are correct. It should be possible to do something similar on the serializer field to what the form field does with `StatusField.get_limit_choices_to()` - looks like creating a custom `get_queryset()` method on `StatusSerializerField` would be the equivalent approach.

File: nautobot/dcim/tests/test_api.py [hunk 1]
```diff
@@ -1485,6 +1485,7 @@ class CableTest(APIViewTestCases.APIViewTestCase):
         "length": 100,
         "length_unit": "m",
     }
+    choices = ["termination_a_type", "termination_b_type", "type", "status", "length_unit"]
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_api.py:1488]: So `choices` is a list of which fields should have choices, not what the values should be for those choices? `choices` is a bit of a vague name to me, perhaps something like `choices_fields` might be clearer?

File: nautobot/utilities/testing/api.py [hunk 2]
```diff
@@ -233,6 +234,27 @@ def test_options_objects_returns_display_and_value(self):
...
                     self.assertIn("value", choice, f"A choice in {field} is missing the value key")
 
+        @override_settings(EXEMPT_VIEW_PERMISSIONS=[])
+        def test_options_returns_expected_choices(self):
+            """
+            Make an OPTIONS request for a list endpoint and validate choices match expected choices.
+            """
+            # Skip test if test case does not supply a list of choices to test against.
+            if not self.choices:
+                self.skipTest("choices was not provided by test case.")
```
Comment (glennmatthews) [inline nautobot/utilities/testing/api.py:244]: Would it instead make sense to validate that *no* OPTIONS choices are returned in this case? I recognize that this might require adding `choices` declarations to a bunch of existing test cases for correctness but that would probably be a good thing to do regardless.

File: nautobot/core/api/metadata.py [hunk 3]
```diff
@@ -65,10 +65,15 @@ def get_field_info(self, field):
...
 
 
-class StatusFieldMetadata(BulkOperationMetadata):
+class StatusFieldMetadata(ContentTypeMetadata):
     """Emit `Status` serializer fields as a choice enum."""
 
     def get_field_info(self, field):
+        # Keep backwards compatible behavior if serializer also has `ContentTypeField` fields specified
+        if isinstance(field, ContentTypeField):
+            return super().get_field_info(field)
```
Comment (glennmatthews) [inline nautobot/core/api/metadata.py:74]: Is this addition actually necessary? We already call `super().get_field_info(field)` below and eventually return it.

**Review by glennmatthews — COMMENTED** [review]
Thank you!

File: nautobot/core/api/metadata.py (hunk 3 above)
Comment (FragmentedPacket) [reply nautobot/core/api/metadata.py:74]: I was thinking of just returning right away rather than hitting the logic below, but it will be removed in an upcoming commit.

**Comment by FragmentedPacket** [comment]
This should be ready for review now with applicable tests and a few minor updates to properly return `content_types` that are serialized as `ContentTypeField`.

**Review by glennmatthews — APPROVED** [review]
Looks great!

**Review by jathanism — APPROVED** [review]
Elegant decision just making `StatusFieldMetadata` a subclass of `ContentTypeMetadata`!
